bakery/slice.go

```diff
@@ -0,0 +1,127 @@
+package bakery
+
+import (
+	"golang.org/x/net/context"
+	"fmt"
+	"time"
+
+	errgo "gopkg.in/errgo.v1"
+	"gopkg.in/macaroon-bakery.v2-unstable/bakery/checkers"
+	macaroon "gopkg.in/macaroon.v2-unstable"
+)
+
+// Slice holds a slice of unbound macaroons.
+type Slice []*Macaroon
+
+// Bind prepares the macaroon slice for use in a request. This must be
+// done before presenting the macaroons to a service for use as
+// authorization tokens. The result will only be valid
+// if s contains discharge macaroons for all third party
+// caveats.
+//
+// All the macaroons in the returned slice will be copies
+// of this in s, not references.
+func (s Slice) Bind() macaroon.Slice {
+	if len(s) == 0 {
+		return nil
+	}
+	ms := make(macaroon.Slice, len(s))
+	ms[0] = s[0].M().Clone()
+	rootSig := ms[0].Signature()
+	for i, m := range s[1:] {
+		m1 := m.M().Clone()
+		m1.Bind(rootSig)
+		ms[i+1] = m1
+	}
+	return ms
+}
+
+// Purge returns a new slice holding all macaroons in s
+// that expire after the given time.
+func (ms Slice) Purge(t time.Time) Slice {
```

Comment:
This should probably be a little more intelligent. if Slice[0] has expired and gets purged doesn't it make the Slice somewhat meaningless, should it perhaps throw everything away then?

Reply:
Excellent point - will do.

Reply:
Done